### ISSUE TITLE
networking.k8s.io/ingress.v1beta1 sticky sessions support

### ISSUE THREAD

**Issue by prateekjainaa** [projectcontour/contour#2323]
**Please describe the problem you have**
We have an application which is comprised of close to 80 different services/applications, all exposed under **single FQDN**. Now, some of them have requirement for supporting 'sticky sessions'. It employs TLS termination for security. One of the primary requirement on this application is, it should allow addition and removal of services from deployment at runtime. This is achieved via use of Ingress objects. 

As we are working to move deployment to use contour. This is cropping up as issue because contour doesn't support **sticky sessions** via **Ingress** objects. It only supports it via HTTPProxy object. But we cannot use HTTPProxy object as it doesn't allow, its objects to be taken out and deploy on the fly. 

It would be nice to have support from contour for sticky sessions using Ingress objects. May be via adding another annotation on Ingress.  

_On a separate note_, do you have any plans to support removal and addition of HTTPProxy objects to cluster at runtime (similar to Ingress objects) without manual intervention? Once we get this feature from contour, it will enable deployments to get rid of Ingress objects.

**Comment by youngnick** [comment]
I'm sorry, I'm not sure what you mean by "removal and addition of HTTPProxy obejcts to cluster at runtime".

All objects that Contour handles are able to be added and removed at any time, and Contour will reflect those changes into the Envoy config. That includes both Ingress and HTTPProxy. If you are seeing a problem with HTTPProxy config not being reflected in Envoy, that is a bug.

**Comment by youngnick** [comment]
I'd also be interested to see if there's any documentation that gave you the impression that HTTPProxy objects would behave in this way, as if there is something, it should be fixed.

**Comment by jpeach** [comment]
IIUC, @prateekjainaa is looking for the Ingress behavior where a services don't need to be explicitly attached to a vhost object (i.e. so the 80 microservice deployments don't all need to update the root HTTPProxy). Depending on how the services and the corresponding teams are organized, HTTPProxy inclusion might be able to address this. You need to set up some static includes from the root proxy down to some leaf proxy that each team can operate.

**Comment by prateekjainaa** [comment]
> IIUC, @prateekjainaa is looking for the Ingress behavior where a services don't need to be explicitly attached to a vhost object (i.e. so the 80 microservice deployments don't all need to update the root HTTPProxy). Depending on how the services and the corresponding teams are organized, HTTPProxy inclusion might be able to address this. You need to set up some static includes from the root proxy down to some leaf proxy that each team can operate.

@jpeach, yes your understanding is correct. But when using Ingress, we dont have to do any static includes. It should not matter if we add or remove applications from cluster. HTTPProxy object should handle it by updating root/central proxy object. We have tested static includes with contour version 1.0 and any time we remove child HTTPProxy object, it gave us error, leaving entire proxy object in invalid state. 

IMHO, it is one of the fundamental requirements from microservices architecture that even if certain services are unavailable (due to error or undeployment); system should be functional.

Let me know in case I am missing something here or something has changed in recent releases.

**Comment by jpeach** [comment]
xref #2206 
xref #2141

**Comment by davecheney** [comment]
Hello,

Thank you for raising this issue. A few things.

1. I've updated the title of the issue to reflect what I think is the main feature request being asked for -- sticky session support for networking.k8s.io/ingress.v1beta1 objects. If this is not correct, please let me know
2. You've made some requests for other things in this issue. As an observation asking for more than one thing in an issue generally results in none of them being done because the discussion never converges.
3. @youngnick has asked several questions about how various application teams collaborating the same vhost work in your cluster. It would be great if you could give us some more specifics about your use case.

Lastly let's talk about stick session support that Contour offers. We call it session affinity, whereby Envoy will assign a cookie (if not present) to each request destined for a route using session affinity as its load balancing strategy. You should read the section (sorry I cannot link to it directly) here, https://projectcontour.io/docs/v1.2.1/httpproxy/

Session affinity is not offered on the k8s ingress object at the moment for the simple reason that the Ingress object does not provide a place to nominate a per route load balancing strategy. To implement this we would have to add it as an annotation on the Ingress object. This could take two forms

1. An annotation that forces all routes on a specific Ingress to use session affinity as their load balancing strategy. Something like `projectcontour.io/loadBalancerPolicy: "Cookie"`.
2. An annotation that nominates specific routes in an ingress object use session affinity as their load balancing strategy. Something like `projectcontour.io/sessionAffinityRoutes: "/foo,/bar,/baz"`. This is similar to the way the `projectcontour.io/websocket-routes` annotation works; ref https://projectcontour.io/docs/v1.2.1/annotations/

**Comment by prateekjainaa** [comment]
> I'm sorry, I'm not sure what you mean by "removal and addition of HTTPProxy obejcts to cluster at runtime".
> 
> All objects that Contour handles are able to be added and removed at any time, and Contour will reflect those changes into the Envoy config. That includes both Ingress and HTTPProxy. If you are seeing a problem with HTTPProxy config not being reflected in Envoy, that is a bug.

@youngnick , lets consider this by example,

--> We have one single namespace lets say, ns1
--> We have 80 applications deployed in this namespace. They are interdependent on each other for some or other functionality.
--> A set of applications can be removed or added into cluster, anytime.

Now, consider the scenario of one HTTPProxy objects have configurations about other like static includes. Consider the complexity of the root HTTPProxy object of application. To add, this product is deployed to 100's of times in varous locations with different set of applications/services in them. 

Now, with Ingress objects all these issues go away because; all objects are independent of each other (from deployment perspective). We can take out any application any time and its controller will detect and update itself. We were asking for the same behavior. But as @davecheney commented, we will open another issue for it. Let this issue be dedicated for session affinity.

**Comment by prateekjainaa** [comment]
> Hello,
> 
> Thank you for raising this issue. A few things.
> 
> 1. I've updated the title of the issue to reflect what I think is the main feature request being asked for -- sticky session support for networking.k8s.io/ingress.v1beta1 objects. If this is not correct, please let me know
> 2. You've made some requests for other things in this issue. As an observation asking for more than one thing in an issue generally results in none of them being done because the discussion never converges.
> 3. @youngnick has asked several questions about how various application teams collaborating the same vhost work in your cluster. It would be great if you could give us some more specifics about your use case.
> 
> Lastly let's talk about stick session support that Contour offers. We call it session affinity, whereby Envoy will assign a cookie (if not present) to each request destined for a route using session affinity as its load balancing strategy. You should read the section (sorry I cannot link to it directly) here, https://projectcontour.io/docs/v1.2.1/httpproxy/
> 
> Session affinity is not offered on the k8s ingress object at the moment for the simple reason that the Ingress object does not provide a place to nominate a per route load balancing strategy. To implement this we would have to add it as an annotation on the Ingress object. This could take two forms
> 
> 1. An annotation that forces all routes on a specific Ingress to use session affinity as their load balancing strategy. Something like `projectcontour.io/loadBalancerPolicy: "Cookie"`.
> 2. An annotation that nominates specific routes in an ingress object use session affinity as their load balancing strategy. Something like `projectcontour.io/sessionAffinityRoutes: "/foo,/bar,/baz"`. This is similar to the way the `projectcontour.io/websocket-routes` annotation works; ref https://projectcontour.io/docs/v1.2.1/annotations/

@davecheney second option given by you looks more flexible. Can we expect that in next release?

also, we have earlier replied to query from @youngnick .  Let us know if there is something else too required from our side.

**Comment by prateekjainaa** [comment]
@davecheney, @michmike any updates on this issue?

**Comment by davecheney** [comment]
@youngnick and @michmike are the decision makers here. I encourage you to join the next community meeting to discuss this https://projectcontour.io/community/

**Comment by youngnick** [comment]
Currently this ticket is not prioritised, so we will need to prioritise it before we can give any estimates about date.

From what I can see, the todo is:
- add a new annotation `projectcontour.io/sessionAffinityRoutes: "/foo,/bar,/baz"`, similar to how the `websocket-routes` annotation works.

It doesn't seem like a large amount of work, so I would hope we should be able to get it prioritised soon.

**Comment by prateekjainaa** [comment]
@youngnick this ticket still tagged as, waiting-for-info. Is there any information pending on our side?

**Comment by youngnick** [comment]
Oh, sorry, I failed to update the labels. Updated now, thanks.

**Comment by prateekjainaa** [comment]
Hi @youngnick/ @davecheney ,

 Did you guys get a chance to look at this issue? 

Regards,
Prateek

**Comment by youngnick** [comment]
Hi @prateekjainaa,

We have not had a chance to look at this issue yet, as you can see it's still in Parking Lot 1 on our Project Board. This means we'll try and get to it soon, but I can't make any guarantees about when that will be, sorry. We have some big refactors happening at the moment, but I'll raise your item with the maintainers and see if anyone can take a look.

In addition, PRs are always welcomed. :)

Lastly, `@davecheney` has left the Contour team and so is not going to be able to help with this issue in the future.

**Comment by prateekjainaa** [comment]
Hi @youngnick,
  How are you? Regarding your suggestion on PR, can I expect any help like guidance on what all changes are expected? As we dont understand this code fully. I quickly glanced current code and can see annotations.go , envoy.go (excluding tests and documentation) as probable targets.  

Regards,
Prateek

**Comment by youngnick** [comment]
Thanks for your interest @prateekjainaa! Yes, I would expect changes in `internal/annotations` for sure, to check for the annotation, and then in `internal/dag` to actually implement them. In this case, you can use `WebsocketRoutes` as a bit of a pointer to the patterns for parsing the annotations, then if you set the `LoadBalancerPolicy` to `Cookie` for the referenced routes, you should get the output you want.

I would expect a PR to include unit tests in any package you need to touch, and a feature test in `internal/featuretests` - `internal/featuretests/websocket_test.go` is again a good place to start.

Please feel free to open an early draft PR if you're not sure about your approach, and I'll help shepherd your PR.

**Comment by xaleeks** [comment]
With the direction away from reliance on annotation, I think this is probably a won’t do at this point.  We have a very robust solution for session affinity through HTTPProxy and I think it will be part of the fast maturing Gateway API project in the future as well.  I’ll leave it open for a bit and close it if there are no objections.

**Comment by xaleeks** [comment]
closed by accident

**Comment by sunjayBhatia** [comment]
seems closable?

**Comment by youngnick** [comment]
Agreed. This is wontfix for now.
